### PR TITLE
Add a callback to swift::reflection::MemoryReader that allows LLDB to… 

### PR DESCRIPTION
**Pull request by adrian-prantl**
Add a callback to swift::reflection::MemoryReader that allows LLDB to provide

swift::reflection::TypeInfo for (Clang-)imported non-Objective-C types. This is
needed to reflect on the size mixed Swift / Clang types.

### CODE REVIEW

**Comment by adrian-prantl** [comment]
@augusto2112 For some reason github did not let me request a review from you.

**Comment by adrian-prantl** [comment]
Some more context: I'm working on a `TypeRef`-based Swift context for LLDB. Because we do not have reflection metadata for pure C types in Swift, reflection cannot compute `TypeInfo` for `NominalTypeRef`s for those types. By providing this callback, LLDB can supply this information for DWARF, and reflection can compute `TypeInfo`s for mixed Swift/C types.

**Comment by slavapestov** [comment]
For pure C types that are referenced from code, we emit BuiltinTypeDescriptors in IRGen. I'm guessing here you need the callback because expression evaluation can pull in arbitrary C types that are not referenced from code? If so this approach sounds great.

I have reservations about adding this to MemoryReader since it's not really about reading memory. Maybe we need a second callback interface, or would that be too awkward to plumb through?

File: include/swift/Reflection/ReflectionContext.h [hunk 1]
```diff
@@ -623,7 +623,7 @@ class ReflectionContext
 
         // Perform layout
         if (start)
-          TI = TC.getClassInstanceTypeInfo(TR, *start);
+          TI = TC.getClassInstanceTypeInfo(TR, *start, getReader());
```
Comment (slavapestov) [inline include/swift/Reflection/ReflectionContext.h:626]: This is why I think a different callback would be nice. In theory it should be possible to look up TypeInfos without a remote process attached; and you could still have a callback to get C type info.

File: stdlib/public/Reflection/TypeLowering.cpp [hunk 2]
```diff
@@ -1395,13 +1399,13 @@ TypeConverter::getThinFunctionTypeInfo() {
...
 const TypeInfo *
-TypeConverter::getThickFunctionTypeInfo() {
+TypeConverter::getThickFunctionTypeInfo(remote::MemoryReader &reader) {
   if (ThickFunctionTI != nullptr)
     return ThickFunctionTI;
 
   RecordTypeInfoBuilder builder(*this, RecordKind::ThickFunction);
-  builder.addField("function", getThinFunctionTypeRef());
-  builder.addField("context", getNativeObjectTypeRef());
+  builder.addField("function", getThinFunctionTypeRef(), reader);
```
Comment (slavapestov) [inline stdlib/public/Reflection/TypeLowering.cpp:1407]: This shouldn't need a reader. A thick function is just a pair of pointers as far as RemoteMirrors are concerned.

File: stdlib/public/Reflection/TypeLowering.cpp [hunk 3]
```diff
@@ -1176,12 +1180,12 @@ class ExistentialTypeInfoBuilder {
       // Class existentials consist of a single retainable pointer
       // followed by witness tables.
       if (Refcounting == ReferenceCounting::Unknown)
-        builder.addField("object", TC.getUnknownObjectTypeRef());
+        builder.addField("object", TC.getUnknownObjectTypeRef(), reader);
```
Comment (slavapestov) [inline stdlib/public/Reflection/TypeLowering.cpp:1183]: None of these should need to read imported TypeInfo either

File: stdlib/public/Reflection/TypeLowering.cpp [hunk 4]
```diff
@@ -1034,7 +1034,11 @@ class ExistentialTypeInfoBuilder {
           ++WitnessTableCount;
 
           if (auto *Superclass = TC.getBuilder().lookupSuperclass(P)) {
-            auto *SuperclassTI = TC.getTypeInfo(Superclass);
+            // ObjC class info should be available in the metadata, so it's safe
+            // to instantiate a dummy reader here. This helps preserving the
+            // layering.
+            remote::MemoryReader dummyReader;
+            auto *SuperclassTI = TC.getTypeInfo(Superclass, dummyReader);
```
Comment (slavapestov) [inline stdlib/public/Reflection/TypeLowering.cpp:1041]: Could the reader parameter to getTypeInfo() be a pointer instead, with handing of the null case?

Comment (augusto2112) [reply stdlib/public/Reflection/TypeLowering.cpp:1041]: Maybe llvm:Optional to make it clear the reader is... well... optional?

Comment (augusto2112) [reply stdlib/public/Reflection/TypeLowering.cpp:1041]: I believe llvm::Optional also has the benefit of being able to keep `queryDataLayout`, `getSymbolAddress` and `readString` deleted in the base class.

**Comment by augusto2112** [comment]
> @augusto2112 For some reason github did not let me request a review from you.

Probably because I'm not a project member (that's why I have to @ people when I want a review).

**Comment by adrian-prantl** [comment]
> For pure C types that are referenced from code, we emit BuiltinTypeDescriptors in IRGen. I'm guessing here you need the callback because expression evaluation can pull in arbitrary C types that are not referenced from code? If so this approach sounds great.

Interesting! The case I primarily debugged this with was using a type referenced in the code, but perhaps the example program was too simple to trigger emission of the metadata. I'll look into that — but it also seems like this could be useful even if that were a bug because of the pulling-extra-types-in scenario you mentioned.

> I have reservations about adding this to MemoryReader since it's not really about reading memory. Maybe we need a second callback interface, or would that be too awkward to plumb through?

That sounds totally reasonable, I'll take a look at how well that fit.

**Comment by adrian-prantl** [comment]
I inspected the LLVM IR generated for my testcase and there is no mention of the imported C type outside of debug info. When adding a `print()` of a variable of that type to the program, this changes. From that I'm concluding that the compiler will only emit the type metadata when there is an actual dependency on it. So it looks like I'm going to need this mechanism regardless.

**Comment by slavapestov** [comment]
I see, you're right. The compiler only emits the builtin descriptor when the type appears as the field type of another type, because the original use case for RemoteMirrors was to enable object graph debugging. It's reasonable to pull in additional type info for the types of local variables and such.

**Comment by adrian-prantl** [comment]
I have now created a new `swift::remote::TypeInfoProvider` interface that is to be passed in on a per-function basis rather than owned by the `TypeConverter`. This actually helps removing some ugly statefulness in the LLDB patch using this.

**Comment by adrian-prantl** [comment]
@slavapestov The LLDB side of this is ready now. Can you check if I addressed all your concerns?

**Review by slavapestov — COMMENTED** [review]
I only have one last suggestion. I think we should stick the TypeInfoProvider inside the TypeConverter, so that it doesn't have to be passed around. What do you think?

**Comment by adrian-prantl** [comment]
>  I think we should stick the TypeInfoProvider inside the TypeConverter, so that it doesn't have to be passed around.

My understanding is that `swift::reflection::ReflectionContext` owns exactly one `Builder`, which owns exactly one `TypeConverter`. LLDB, however, has one `TypeInfoProvider` per `lldb::Module` (≈ dylib) to localize type lookup, and one `ReflectionContext` per process. That's why it's passed in explicitly.

**Comment by slavapestov** [comment]
I didn't realize we had one TypeInfoProvider per Module. In that case, I think it's fine to do what you're doing now (but we could also investigate having one TypeConverter per module or something?).

**Comment by shahmishal** [comment]
@adrian-prantl This PR need to be updated to target master branch form master-rebranch.

**Comment by adrian-prantl** [comment]
@swift-ci test

**Comment by adrian-prantl** [comment]
@swift-ci test

**Comment by adrian-prantl** [comment]
@swift-ci smoke test

**Comment by adrian-prantl** [comment]
@swift-ci smoke test

**Comment by adrian-prantl** [comment]
@swift-ci smoke test

**Comment by adrian-prantl** [comment]
test with https://github.com/apple/llvm-project/pull/1636
@swift-ci test

**Comment by adrian-prantl** [comment]
test with https://github.com/apple/llvm-project/pull/1636
@swift-ci smoke test

**Comment by adrian-prantl** [comment]
test with https://github.com/apple/llvm-project/pull/1636
@swift-ci test

**Comment by adrian-prantl** [comment]
test with https://github.com/apple/llvm-project/pull/1636
@swift-ci smoke test

**Comment by adrian-prantl** [comment]
test with https://github.com/apple/llvm-project/pull/1636
@swift-ci test

**Comment by adrian-prantl** [comment]
test with https://github.com/apple/llvm-project/pull/1636
@swift-ci test

**Comment by adrian-prantl** [comment]
test with https://github.com/apple/llvm-project/pull/1636
@swift-ci smoke test

**Comment by compnerd** [comment]
@adrian-prantl this seems to have caused a regression on the Windows builds :-(

**Pull request merged**